packages/cosmic-swingset/lib/ag-solo/vats/lib-wallet.js

```diff
@@ -341,37 +346,115 @@ export async function makeWallet({
+      return Object.fromEntries(
+        Object.entries(amountKeywordRecord).map(
+          ([keyword, { pursePetname, extent }]) => {
+            const purse = getPurse(pursePetname);
+            purseKeywordRecord[keyword] = purse;
+            const brand = purseToBrand.get(purse);
+            const amount = { brand, extent };
+            return [keyword, amount];
+          },
+        ),
+      );
+    };
+
+    const proposal = {
+      want: compile(want),
+      give: compile(give),
+      exit,
+    };
+
+    return { proposal, purseKeywordRecord };
+  };
+
+  const display = value => fillInSlots(dehydrate(harden(value)));
+
+  const displayProposal = proposalTemplate => {
+    const {
+      want = {},
+      give = {},
+      exit = { onDemand: null },
+    } = proposalTemplate;
+    const compile = pursePetnameExtentKeywordRecord => {
+      return Object.fromEntries(
+        Object.entries(pursePetnameExtentKeywordRecord).map(
+          ([keyword, { pursePetname, extent }]) => {
+            const purse = getPurse(pursePetname);
+            const brand = purseToBrand.get(purse);
+            const amount = { brand, extent };
+            return [keyword, { pursePetname, amount: display(amount) }];
+          },
+        ),
+      );
+    };
+    const proposal = {
+      want: compile(want),
+      give: compile(give),
+      exit,
+    };
+    return proposal;
+  };
+
+  const compileOffer = async offer => {
+    const { inviteHandleBoardId } = offer;
+    const { proposal, purseKeywordRecord } = compileProposal(
+      offer.proposalTemplate,
+    );
+
+    // Find invite in wallet and withdraw
+    const defaultInvitePurse = getPurse(ZOE_INVITE_PURSE_PETNAME);
+    const { extent: inviteExtentElems } = await E(
+      defaultInvitePurse,
+    ).getCurrentAmount();
+    const inviteHandle = await E(board).getValue(inviteHandleBoardId);
+    const matchInvite = element => element.handle === inviteHandle;
+    const inviteBrand = purseToBrand.get(defaultInvitePurse);
+    const { amountMath: inviteAmountMath } = brandTable.get(inviteBrand);
+    const inviteAmount = inviteAmountMath.make(
+      harden([inviteExtentElems.find(matchInvite)]),
+    );
+    const inviteP = E(defaultInvitePurse).withdraw(inviteAmount);
+
+    return { proposal, inviteP, purseKeywordRecord };
+  };
+
+  async function addOffer(rawOffer, requestContext = { origin: 'unknown' }) {
+    const {
+      id: rawId,
+      instanceHandleBoardId,
+      installationHandleBoardId,
+      proposalTemplate,
+    } = rawOffer;
     const id = `${requestContext.origin}#${rawId}`;
-    const offer = {
+    const instanceHandle = await E(board).getValue(instanceHandleBoardId);
+    const installationHandle = await E(board).getValue(
+      installationHandleBoardId,
+    );
+    const offer = harden({
       ...rawOffer,
       id,
       requestContext,
       status: undefined,
-    };
+      instancePetname: display(instanceHandle).petname,
```

Comment:
This is okay if we're assuming petnames are static.  Saying that, maybe it's a good idea to document that assumption in a comment somewhere.

Otherwise, I would have expected the display of the instance/installation/proposal to be postponed to the time it's actually needed.

Reply:
Can you say more about when the display would happen, compared to here? when updateInboxState is called?

Reply:
Ah, right. 🙂 

I think what I had in mind was for some future in which we also update state for petname changes, but that doesn't make sense right now.

Reply:
Ok, np, I feel like much of this will change in the near future, so happy to revisit!